### PR TITLE
Change cluster status default date type

### PR DESCRIPTION
**Pull request by GGP1**
|Related issue|
|---|
|Closes https://github.com/wazuh/wazuh/issues/16995 |

## Description

Changed the default date string from a `datetime` at the unix epoch to a string `"n/a"`.

## Tests

Cluster unit and integration tests were executed successfully.

```console
(venv2) gasti@pop-os:~/work/wazuh$ python3 -m pytest framework/wazuh/core/cluster/tests/test_master.py
================================================================ test session starts =================================================================
platform linux -- Python 3.9.16, pytest-5.4.3, py-1.11.0, pluggy-0.13.1
rootdir: /home/gasti/work/wazuh/framework
plugins: aiohttp-0.3.0, tavern-1.2.2, trio-0.7.0, html-2.1.1, asyncio-0.15.1, metadata-2.0.4
collected 47 items                                                                                                                                   

framework/wazuh/core/cluster/tests/test_master.py ...............................................                                              [100%]

=========================================================== 47 passed, 5 warnings in 0.55s ===========================================================
```

```console
(venv2) gasti@pop-os:~/work/wazuh/api/test/integration$ pytest -vv test_cluster_endpoints.tavern.yaml
================================================================ test session starts =================================================================
platform linux -- Python 3.9.16, pytest-5.4.3, py-1.11.0, pluggy-0.13.1 -- /home/gasti/work/wazuh/venv2/bin/python
cachedir: .pytest_cache
metadata: {'Python': '3.9.16', 'Platform': 'Linux-6.2.6-76060206-generic-x86_64-with-glibc2.35', 'Packages': {'pytest': '5.4.3', 'pluggy': '0.13.1'}, 'Plugins': {'aiohttp': '0.3.0', 'tavern': '1.2.2', 'trio': '0.7.0', 'html': '2.1.1', 'asyncio': '0.15.1', 'metadata': '2.0.4'}}
rootdir: /home/gasti/work/wazuh/api/test/integration, inifile: pytest.ini
plugins: aiohttp-0.3.0, tavern-1.2.2, trio-0.7.0, html-2.1.1, asyncio-0.15.1, metadata-2.0.4
collected 49 items                                                                                                                                   

test_cluster_endpoints.tavern.yaml::GET /cluster/local/config PASSED                                                                           [  2%]
test_cluster_endpoints.tavern.yaml::GET /cluster/healthcheck PASSED                                                                            [  4%]
test_cluster_endpoints.tavern.yaml::GET /cluster/ruleset/synchronization PASSED                                                                [  6%]
test_cluster_endpoints.tavern.yaml::GET /cluster/local/info PASSED                                                                             [  8%]
test_cluster_endpoints.tavern.yaml::GET /cluster/nodes PASSED                                                                                  [ 10%]
test_cluster_endpoints.tavern.yaml::GET /cluster/nodes[master-node] PASSED                                                                     [ 12%]
test_cluster_endpoints.tavern.yaml::GET /cluster/nodes[worker1] PASSED                                                                         [ 14%]
test_cluster_endpoints.tavern.yaml::GET /cluster/nodes[worker2] PASSED                                                                         [ 16%]
test_cluster_endpoints.tavern.yaml::GET /cluster/nodes select[ip] PASSED                                                                       [ 18%]
test_cluster_endpoints.tavern.yaml::GET /cluster/nodes select[name] PASSED                                                                     [ 20%]
test_cluster_endpoints.tavern.yaml::GET /cluster/nodes select[type] PASSED                                                                     [ 22%]
test_cluster_endpoints.tavern.yaml::GET /cluster/nodes select[version] PASSED                                                                  [ 24%]
test_cluster_endpoints.tavern.yaml::GET /cluster/status PASSED                                                                                 [ 26%]
test_cluster_endpoints.tavern.yaml::GET /cluster/{node_id}/configuration PASSED                                                                [ 28%]
test_cluster_endpoints.tavern.yaml::GET /cluster/configuration/validation PASSED                                                               [ 30%]
test_cluster_endpoints.tavern.yaml::GET /cluster/api/config PASSED                                                                             [ 32%]
test_cluster_endpoints.tavern.yaml::GET /cluster/{node_id}/configuration/{component}/{configuration} PASSED                                    [ 34%]
test_cluster_endpoints.tavern.yaml::GET /cluster/{node_id}/info[master-node] PASSED                                                            [ 36%]
test_cluster_endpoints.tavern.yaml::GET /cluster/{node_id}/info[worker1] PASSED                                                                [ 38%]
test_cluster_endpoints.tavern.yaml::GET /cluster/{node_id}/info[worker2] PASSED                                                                [ 40%]
test_cluster_endpoints.tavern.yaml::GET /cluster/{node_id}/logs[master-node] PASSED                                                            [ 42%]
test_cluster_endpoints.tavern.yaml::GET /cluster/{node_id}/logs[worker1] PASSED                                                                [ 44%]
test_cluster_endpoints.tavern.yaml::GET /cluster/{node_id}/logs[worker2] PASSED                                                                [ 46%]
test_cluster_endpoints.tavern.yaml::GET /cluster/{node_id}/logs/summary[worker1] PASSED                                                        [ 48%]
test_cluster_endpoints.tavern.yaml::GET /cluster/{node_id}/logs/summary[worker2] PASSED                                                        [ 51%]
test_cluster_endpoints.tavern.yaml::GET /cluster/{node_id}/daemons/stats[master-node] PASSED                                                   [ 53%]
test_cluster_endpoints.tavern.yaml::GET /cluster/{node_id}/daemons/stats[worker1] PASSED                                                       [ 55%]
test_cluster_endpoints.tavern.yaml::GET /cluster/{node_id}/daemons/stats[worker2] PASSED                                                       [ 57%]
test_cluster_endpoints.tavern.yaml::GET /cluster/{node_id}/stats[master-node] PASSED                                                           [ 59%]
test_cluster_endpoints.tavern.yaml::GET /cluster/{node_id}/stats[worker1] PASSED                                                               [ 61%]
test_cluster_endpoints.tavern.yaml::GET /cluster/{node_id}/stats[worker2] PASSED                                                               [ 63%]
test_cluster_endpoints.tavern.yaml::GET /cluster/wrong_node/stats PASSED                                                                       [ 65%]
test_cluster_endpoints.tavern.yaml::GET /cluster/{node_id}/stats/analysisd[master-node] PASSED                                                 [ 67%]
test_cluster_endpoints.tavern.yaml::GET /cluster/{node_id}/stats/analysisd[worker1] PASSED                                                     [ 69%]
test_cluster_endpoints.tavern.yaml::GET /cluster/{node_id}/stats/analysisd[worker2] PASSED                                                     [ 71%]
test_cluster_endpoints.tavern.yaml::GET /cluster/{node_id}/stats/hourly[master-node] PASSED                                                    [ 73%]
test_cluster_endpoints.tavern.yaml::GET /cluster/{node_id}/stats/hourly[worker1] PASSED                                                        [ 75%]
test_cluster_endpoints.tavern.yaml::GET /cluster/{node_id}/stats/hourly[worker2] PASSED                                                        [ 77%]
test_cluster_endpoints.tavern.yaml::GET /cluster/{node_id}/stats/remoted[master-node] PASSED                                                   [ 79%]
test_cluster_endpoints.tavern.yaml::GET /cluster/{node_id}/stats/remoted[worker1] PASSED                                                       [ 81%]
test_cluster_endpoints.tavern.yaml::GET /cluster/{node_id}/stats/remoted[worker2] PASSED                                                       [ 83%]
test_cluster_endpoints.tavern.yaml::GET /cluster/{node_id}/stats/weekly[master-node] PASSED                                                    [ 85%]
test_cluster_endpoints.tavern.yaml::GET /cluster/{node_id}/stats/weekly[worker1] PASSED                                                        [ 87%]
test_cluster_endpoints.tavern.yaml::GET /cluster/{node_id}/stats/weekly[worker2] PASSED                                                        [ 89%]
test_cluster_endpoints.tavern.yaml::GET /cluster/{node_id}/status[master-node] PASSED                                                          [ 91%]
test_cluster_endpoints.tavern.yaml::GET /cluster/{node_id}/status[worker1] PASSED                                                              [ 93%]
test_cluster_endpoints.tavern.yaml::GET /cluster/{node_id}/status[worker2] PASSED                                                              [ 95%]
test_cluster_endpoints.tavern.yaml::PUT /cluster/restart PASSED                                                                                [ 97%]
test_cluster_endpoints.tavern.yaml::PUT /cluster/{node_id}/configuration PASSED                                                                [100%]

==================================================== 49 passed, 62 warnings in 453.31s (0:07:33) =====================================================

### CODE REVIEW

**Review by EduLeon12 — DISMISSED** [review]
All Tests passed properly, development was well executed and the issue was resolved.

LGTM !!

**Review by mhamra — CHANGES_REQUESTED** [review]
Review:
- Unit and Integration tests passed ✔️ 
- Code review ✔️ 

Recommendation:
- Add the last empty line to the "master.py" file to follow PEP standards.

**Comment by GGP1** [comment]
> Recommendation:
> 
>     * Add the last empty line to the "master.py" file to follow PEP standards.

Rebasing to master added the line at the end of the file.

**Comment by GGP1** [comment]
> Recommendation:
> 
>     * Add the last empty line to the "master.py" file to follow PEP standards.

Rebase to master added the line at the end of the file.

**Review by mhamra — APPROVED** [review]
Gasti, 
After rebasing, LGTM!

File: framework/wazuh/core/cluster/master.py [hunk 1]
```diff
@@ -266,14 +267,20 @@ def process_request(self, command: bytes, data: bytes) -> Tuple[bytes, bytes]:
             logger = self.task_loggers['Agent-groups send']
             start_time = self.send_agent_groups_status['date_start']
             if isinstance(start_time, str):
-                start_time = datetime.strptime(start_time, DECIMALS_DATE_FORMAT)
+                if start_time == DEFAULT_DATE:
+                    start_time = datetime(1970, 1, 1, 0, 0, tzinfo=timezone.utc)
+                else:
+                    start_time = datetime.strptime(start_time, DECIMALS_DATE_FORMAT)
```
Comment (Selutario) [inline framework/wazuh/core/cluster/master.py:273]: It seems that the purpose of the issue has been misunderstood. What is intended is to simplify the current state of the dates in master nodes.

The cluster is a bit complex, but for example in this case, the master cannot receive the `b'syn_w_g_e` command without having previously run the `send_agent_groups_information` method. In said method, a date is already set inside `send_agent_groups_status['date_start']`, so it is not possible for the value of `start_time` to be `DEFAULT_DATE` at this point. Therefore, those `start_time == DEFAULT_DATE` conditions are not needed.

I would expect something similar to this:
```PYTHON
        elif command == b'syn_w_g_e':
            logger = self.task_loggers['Agent-groups send']
            start_time = datetime.strptime(self.send_agent_groups_status['date_start'], DECIMALS_DATE_FORMAT)
            return c_common.end_sending_agent_information(logger, start_time.timestamp(), data.decode())
```

File: framework/wazuh/core/cluster/master.py [hunk 2]
```diff
@@ -292,9 +299,7 @@ def process_request(self, command: bytes, data: bytes) -> Tuple[bytes, bytes]:
             return cmd, json.dumps(res).encode()
         elif command == b'get_health':
             cmd, res = self.get_health(json.loads(data))
-            return cmd, json.dumps(
-                res, default=lambda o: "n/a" if isinstance(o, datetime) and o == utils.get_date_from_timestamp(0) else
-                (o.__str__() if isinstance(o, datetime) else None)).encode()
+            return cmd, json.dumps(res).encode()
```
Comment (Selutario) [inline framework/wazuh/core/cluster/master.py:302]: There is a similar case in `local_server.py`

**Review by Selutario — CHANGES_REQUESTED** [review]

**Review by Selutario — CHANGES_REQUESTED** [review]
I'm getting this error in the `cluster.log`. More changes will be needed than it may seem at first:
```
2023/07/13 16:57:19 INFO: [Worker worker1] [Agent-groups send] Starting.
2023/07/13 16:57:19 ERROR: [Worker worker1] [Main] Unhandled error processing request 'b'syn_w_g_e'': unsupported operand type(s) for -: 'float' and 'datetime.datetime'
Traceback (most recent call last):
  File "/var/ossec/framework/python/lib/python3.9/site-packages/wazuh-4.7.0-py3.9.egg/wazuh/core/cluster/common.py", line 816, in dispatch
    command, payload = self.process_request(command, payload)
  File "/var/ossec/framework/python/lib/python3.9/site-packages/wazuh-4.7.0-py3.9.egg/wazuh/core/cluster/master.py", line 269, in process_request
    return c_common.end_sending_agent_information(logger, start_time, data.decode())
  File "/var/ossec/framework/python/lib/python3.9/site-packages/wazuh-4.7.0-py3.9.egg/wazuh/core/cluster/common.py", line 1628, in end_sending_agent_information
    msg = f"Finished in {(utils.get_utc_now().timestamp() - start_time):.3f}s. Updated {data['updated_chunks']} chunks."
TypeError: unsupported operand type(s) for -: 'float' and 'datetime.datetime'
```

File: framework/wazuh/core/cluster/common.py [hunk 3]
```diff
@@ -1625,7 +1626,9 @@ def end_sending_agent_information(logger, start_time, response) -> Tuple[bytes,
         Response message.
     """
     data = json.loads(response)
-    msg = f"Finished in {(utils.get_utc_now().timestamp() - start_time):.3f}s. Updated {data['updated_chunks']} chunks."
+    msg = f"Finished in {(utils.get_utc_now().timestamp() - start_time.timestamp()):.3f}s. " + \
```
[the remark this answers is not in the document]
Comment (GGP1) [reply framework/wazuh/core/cluster/common.py:1629]: Started a cluster with 1 master, 2 workers and 2 agents. Did not receive any error after the changes
```console
docker-wazuh-agent-1    | Completed.
docker-wazuh-agent2-1   | Completed.
docker-wazuh-master-1   | Completed.
docker-wazuh-worker2-1  | Completed.
docker-wazuh-worker1-1  | Completed.
```

```console
root@wazuh-master:/# /var/ossec/bin/cluster_control -a
ID   NAME          IP          STATUS  VERSION       NODE NAME    
000  wazuh-master  127.0.0.1   active  Wazuh v4.8.0  master-node  
001  042c00921f4f  172.24.0.6  active  Wazuh v4.8.0  worker       
002  de8aa0658595  172.24.0.7  active  Wazuh v4.8.0  master-node
```

I also tested the Integrity sync command and it works as expected

```console
root@wazuh-worker2:/# date; ls -l /var/ossec/etc/rules/
Wed Jul 19 15:56:45 UTC 2023
total 4
-rw-rw---- 1 wazuh wazuh 497 Jul 19 15:56 local_rules.xml
root@wazuh-worker2:/# rm -rf /var/ossec/etc/rules/local_rules.xml
root@wazuh-worker2:/# date; ls -l /var/ossec/etc/rules/
Wed Jul 19 15:56:48 UTC 2023
total 0
root@wazuh-worker2:/# date; ls -l /var/ossec/etc/rules/
Wed Jul 19 15:56:54 UTC 2023
total 4
-rw-rw---- 1 wazuh wazuh 497 Jul 19 15:56 local_rules.xml
```

[the remark this answers is not in the document]
Comment (GGP1) [reply framework/wazuh/core/cluster/common.py:1629]: Started a cluster with 1 master, 2 workers and 2 agents. Did not receive any error after the changes
```console
docker-wazuh-agent-1    | Completed.
docker-wazuh-agent2-1   | Completed.
docker-wazuh-master-1   | Completed.
docker-wazuh-worker2-1  | Completed.
docker-wazuh-worker1-1  | Completed.
```

```console
root@wazuh-master:/# /var/ossec/bin/cluster_control -a
ID   NAME          IP          STATUS  VERSION       NODE NAME    
000  wazuh-master  127.0.0.1   active  Wazuh v4.8.0  master-node  
001  042c00921f4f  172.24.0.6  active  Wazuh v4.8.0  worker       
002  de8aa0658595  172.24.0.7  active  Wazuh v4.8.0  master-node
```

I also tested the Integrity sync command and it works as expected

```console
root@wazuh-worker2:/# date; ls -l /var/ossec/etc/rules/
Wed Jul 19 15:56:45 UTC 2023
total 4
-rw-rw---- 1 wazuh wazuh 497 Jul 19 15:56 local_rules.xml
root@wazuh-worker2:/# rm -rf /var/ossec/etc/rules/local_rules.xml
root@wazuh-worker2:/# date; ls -l /var/ossec/etc/rules/
Wed Jul 19 15:56:48 UTC 2023
total 0
root@wazuh-worker2:/# date; ls -l /var/ossec/etc/rules/
Wed Jul 19 15:56:54 UTC 2023
total 4
-rw-rw---- 1 wazuh wazuh 497 Jul 19 15:56 local_rules.xml
```

**Review by Selutario — APPROVED** [review]
LGTM

**Pull request merged**